src/orbits/keplerian.jl

```diff
@@ -0,0 +1,532 @@
+    a_star = a * M_planet / M_tot
+    a_planet = -a * M_star / M_tot
+
+    # Omega
+    if isnothing(nt.Omega)
+        Omega = 0.0
+    else
+        Omega = nt.Omega
+    end
+    sin_Omega, cos_Omega = sincos(Omega)
+
+    # Eccentricity
+    if isnothing(nt.ecc)
+        ecc = 0.0
+        M_0 = 0.5 * π # TODO: find out why this fails
+        incl_factor_inv = 1.0
+        omega = zero(Omega)
+        cos_omega, sin_omega = zero(omega), oneunit(omega)
+    else
+        ecc = nt.ecc
+        if !isnothing(nt.omega)
+            all( (!isnothing).((nt.cos_omega, nt.sin_omega)) ) && throw(ArgumentError(
+                "Only `ω`, or `cos_ω` and `sin_ω` can be provided"
+            ))
+            omega = nt.omega
+            sin_omega, cos_omega = sincos(nt.omega)
+        elseif all( (!isnothing).((nt.cos_omega, nt.sin_omega)) )
+            cos_omega, sin_omega = nt.cos_omega, nt.sin_omega
+            omega = atan(sin_omega, cos_omega)
+        else
+            throw(ArgumentError("`ω` must also be provided if `ecc` specified"))
+        end
+        E_0 = 2.0 * atan(√(1.0 - ecc) * cos_omega, √(1.0 + ecc) * (1.0 + sin_omega))
+        M_0 = E_0 - ecc * sin(E_0)
+
+        incl_factor_inv  = (1.0 - ecc^2) / (1.0 + ecc * sin_omega)
+    end
+
+    # Jacobian for cos(i) -> b
+    dcosi_db = R_star / a * (1.0 / incl_factor_inv)
+
+    if !isnothing(nt.b)
+        any( (!isnothing).((nt.incl, duration)) ) && throw(ArgumentError(
+            "Only `incl`, `b`, or `duration` can be given"
+        ))
+        b = nt.b
+        cos_incl = dcosi_db * b
+        incl = acos(cos_incl)
+        sin_incl = sin(incl)
+        duration = nt.duration
+    elseif !isnothing(nt.incl)
+        !isnothing(nt.duration) && throw(ArgumentError(
+            "Only `incl`, `b`, or `duration` can be given"
+        ))
+        incl = nt.incl
+        sin_incl, cos_incl = sincos(incl)
+        b = cos_incl / dcosi_db
+        duration = nt.duration
+    elseif !isnothing(nt.duration)
+        duration = nt.duration
+        c = sin(π * duration / (incl_factor_inv) / period)
+        c_sq = c^2
+        ecc_sin_omega = ecc*sin_omega
+        aor = a_planet / R_star
+        b = √(
+            (aor^2 * c_sq - 1.0) /
+            (
+                c_sq * ecc_sin_omega^2 +
+                2.0*c_sq*ecc_sin_omega +
+                c_sq - ecc^4 + 2.0*ecc^2 - 1.0
+            )
+        ) * (1.0 - ecc^2)
+        cos_incl = dcosi_db * b
+        incl = acos(cos_incl)
+        sin_incl = sin(incl)
+    else
+        incl = 0.5*π
+        cos_incl = 0.0
+        sin_incl = 1.0
+        b = 0.0
+        duration = nt.duration
+    end
+
+    # Compute remaining system parameters
+    !(isnothing(nt.t_0) ⊻ isnothing(nt.t_p)) && throw(
+        ArgumentError("Either t₀ or tₚ must be specified")
+    )
+    if isnothing(nt.t_0)
+        t_p = nt.t_p
+        t_0 = t_p + M_0 / n
+    else
+        t_0 = nt.t_0
+        t_p = t_0 - M_0 / n
+    end
```

Comment:
I would federate this out to a helper function, all of these calculations as much as possible. That way we can use re-use them, test them more succinctly, and more importantly they organize the code. You've already got it for a lot of the code, but you can add it here, too

```julia
compute_t0_tp(::Nothing, tp; M0, n) = (tp + M0/n, tp)
compute_t0_tp(t0, ::Nothing; M0, n) = (t0, t0 - M0/n)
```
This code will automatically error if both tp and t0 are provided, or neither, with missing method errors and method ambiguity errors.

If we want to add more verbose error messages (maybe?), we can add
```julia
compute_t0_tp(::Nothing, ::Nothing; kwargs...) = error("Please specify either `t_0` or `t_p`")
compute_t0_tp(t0, tp; kwargs...) = error("Please only specify one of `t_0` or `t_p`")
```

Consider also finding a way to "funcitonize" the relationship between `i`, `duration`, and `b` in the if/else block before this one.

Reply:
Nice, on it! Here's the switch to `compute_t0_tp` 8a7cdeb

* Is there a reason to drop the var names from the method signatures (`x::Nothing` -> `::Nothing`), or is this just a style choice? I've kept them for now
* Is there a preference for `error(...)` over `throw(ArgumentError`(...))? I've just gone with the latter for now

Reply:
Ok, I organized more functions into their own helper file for now 5677b3f